File: java/src/main/java/org/unlaxer/tramli/Pipeline.java
package org.unlaxer.tramli;

import java.util.*;
import java.util.function.Consumer;

/**
 * Build-time verified pipeline — sequential step chain with requires/produces contracts.
 *
 * <pre>
 * var pipeline = Tramli.pipeline("etl")
 *     .initiallyAvailable(RawInput.class)
 *     .step(parse).step(validate).step(save)
 *     .build();
 * FlowContext result = pipeline.execute(Map.of(RawInput.class, data));
 * </pre>
 */
public final class Pipeline {
    private final String name;
    private final List<PipelineStep> steps;
    private final Set<Class<?>> initiallyAvailable;
    private final PipelineDataFlow dataFlow;
    private boolean strictMode;
    private Consumer<LogEntry.Transition> transitionLogger;
    private Consumer<LogEntry.State> stateLogger;
    private Consumer<LogEntry.Error> errorLogger;

    private Pipeline(String name, List<PipelineStep> steps, Set<Class<?>> initiallyAvailable) {
        this.name = name;
        this.steps = List.copyOf(steps);
        this.initiallyAvailable = Set.copyOf(initiallyAvailable);
        this.dataFlow = new PipelineDataFlow(steps, initiallyAvailable);
    }

    public String name() { return name; }
    public PipelineDataFlow dataFlow() { return dataFlow; }

    public void setStrictMode(boolean strict) { this.strictMode = strict; }
    public void setTransitionLogger(Consumer<LogEntry.Transition> l) { this.transitionLogger = l; }
    public void setStateLogger(Consumer<LogEntry.State> l) { this.stateLogger = l; }
    public void setErrorLogger(Consumer<LogEntry.Error> l) { this.errorLogger = l; }
    public void removeAllLoggers() {
        this.transitionLogger = null;
        this.stateLogger = null;
        this.errorLogger = null;
    }

    /**
     * Execute the pipeline. Returns the FlowContext with all produced data.
     * @throws PipelineException if a step fails
     */
    public FlowContext execute(Map<Class<?>, Object> initialData) {
        String flowId = UUID.randomUUID().toString();
        var ctx = new FlowContext(flowId);
        for (var entry : initialData.entrySet()) {
            @SuppressWarnings({"unchecked", "rawtypes"})
            Class key = entry.getKey();
            ctx.put(key, entry.getValue());
        }

        var completed = new ArrayList<String>();
        String prev = "initial";

        for (var step : steps) {
            long stepStart = System.nanoTime();

            // Transition log
            if (transitionLogger != null) {
                transitionLogger.accept(new LogEntry.Transition(flowId, name, prev, step.name(), step.name(), 0));
            }

            // State logger: capture keys before
            Set<Class<?>> keysBefore = stateLogger != null ? new HashSet<>(ctx.snapshot().keySet()) : null;

            try {
                step.process(ctx);
            } catch (Exception e) {
                if (errorLogger != null) {
                    long durationMicros = (System.nanoTime() - stepStart) / 1000;
                    errorLogger.accept(new LogEntry.Error(flowId, name, prev, step.name(), step.name(), e, durationMicros));
                }
                throw new PipelineException(step.name(), completed, ctx, e);
            }

            // strictMode: verify produces
            if (strictMode) {
                for (Class<?> prod : step.produces()) {
                    if (!ctx.has(prod)) {
                        var ex = new FlowException("PRODUCES_VIOLATION",
                                "Step '" + step.name() + "' declares produces " + prod.getSimpleName() + " but did not put it");
                        throw new PipelineException(step.name(), completed, ctx, ex);
                    }
                }
            }

            // State logger: emit new keys
            if (stateLogger != null && keysBefore != null) {
                for (var key : ctx.snapshot().keySet()) {
                    if (!keysBefore.contains(key)) {
                        stateLogger.accept(new LogEntry.State(flowId, name, step.name(), key, ctx.snapshot().get(key)));
                    }
                }
            }

            completed.add(step.name());
            prev = step.name();
        }

        return ctx;
    }

    /**
     * Convert this Pipeline to a PipelineStep for nesting in another Pipeline.
     */
    public PipelineStep asStep() {
        return new PipelineStep() {
            @Override public String name() { return Pipeline.this.name; }
            @Override public Set<Class<?>> requires() { return Pipeline.this.initiallyAvailable; }
            @Override public Set<Class<?>> produces() {
                var all = new LinkedHashSet<Class<?>>();
                for (var s : steps) all.addAll(s.produces());
                return all;
            }
            @Override public void process(FlowContext ctx) {
                // Re-execute steps on the shared context
                for (var s : steps) s.process(ctx);
            }
        };
    }

    // ─── Builder ─────────────────────────────────────────

    public static PipelineBuilder builder(String name) {
        return new PipelineBuilder(name);
    }

    public static final class PipelineBuilder {
        private final String name;
        private final List<PipelineStep> steps = new ArrayList<>();
        private final Set<Class<?>> initiallyAvailable = new LinkedHashSet<>();

        private PipelineBuilder(String name) { this.name = name; }

        public PipelineBuilder initiallyAvailable(Class<?>... types) {
            Collections.addAll(initiallyAvailable, types);
            return this;
        }

        public PipelineBuilder step(PipelineStep step) {
            steps.add(step);
            return this;
        }

        public Pipeline build() {
            // Validate requires/produces chain
            var errors = new ArrayList<String>();
            var available = new LinkedHashSet<>(initiallyAvailable);
            for (var step : steps) {
                for (var req : step.requires()) {
                    if (!available.contains(req)) {
                        errors.add("Step '" + step.name() + "' requires " + req.getSimpleName() + " but it may not be available");
                    }
                }
                available.addAll(step.produces());
            }
            if (!errors.isEmpty()) {
                throw new FlowException("INVALID_PIPELINE",
                        "Pipeline '" + name + "' has " + errors.size() + " error(s):\n  - " + String.join("\n  - ", errors));
            }
            return new Pipeline(name, steps, initiallyAvailable);
        }
    }
}

File: lang/ts/src/pipeline.ts
import { FlowContext } from './flow-context.js';
import { FlowError } from './flow-error.js';
import type { FlowKey } from './flow-key.js';
import type { TransitionLogEntry, StateLogEntry, ErrorLogEntry } from './flow-engine.js';

export interface PipelineStep {
  name: string;
  requires: FlowKey<unknown>[];
  produces: FlowKey<unknown>[];
  process(ctx: FlowContext): Promise<void> | void;
}

export class PipelineException extends FlowError {
  constructor(
    public readonly failedStep: string,
    public readonly completedSteps: string[],
    public readonly context: FlowContext,
    public readonly cause: Error,
  ) {
    super('PIPELINE_STEP_FAILED', `Pipeline step '${failedStep}' failed: ${cause.message}`);
  }
}

export class PipelineDataFlow {
  constructor(
    private readonly steps: PipelineStep[],
    private readonly initiallyAvailable: Set<string>,
  ) {}

  deadData(): Set<string> {
    const allProduced = new Set(this.initiallyAvailable);
    const allConsumed = new Set<string>();
    for (const s of this.steps) {
      for (const r of s.requires) allConsumed.add(r as string);
      for (const p of s.produces) allProduced.add(p as string);
    }
    const dead = new Set<string>();
    for (const p of allProduced) { if (!allConsumed.has(p)) dead.add(p); }
    return dead;
  }

  stepOrder(): string[] { return this.steps.map(s => s.name); }

  availableAfter(stepName: string): Set<string> {
    const available = new Set(this.initiallyAvailable);
    for (const s of this.steps) {
      for (const p of s.produces) available.add(p as string);
      if (s.name === stepName) return available;
    }
    return available;
  }

  toMermaid(): string {
    const lines = ['flowchart LR'];
    for (const s of this.steps) {
      for (const r of s.requires) lines.push(`    ${r} -->|requires| ${s.name}`);
      for (const p of s.produces) lines.push(`    ${s.name} -->|produces| ${p}`);
    }
    return lines.join('\n') + '\n';
  }
}

export class Pipeline {
  readonly name: string;
  private readonly steps: PipelineStep[];
  private readonly _initiallyAvailable: Set<string>;
  private readonly _dataFlow: PipelineDataFlow;
  private strictMode = false;
  private transitionLogger?: (entry: TransitionLogEntry) => void;
  private stateLogger?: (entry: StateLogEntry) => void;
  private errorLogger?: (entry: ErrorLogEntry) => void;

  private constructor(name: string, steps: PipelineStep[], initiallyAvailable: Set<string>) {
    this.name = name;
    this.steps = [...steps];
    this._initiallyAvailable = new Set(initiallyAvailable);
    this._dataFlow = new PipelineDataFlow(steps, initiallyAvailable);
  }

  dataFlow(): PipelineDataFlow { return this._dataFlow; }

  setStrictMode(strict: boolean): void { this.strictMode = strict; }
  setTransitionLogger(l: ((e: TransitionLogEntry) => void) | null): void { this.transitionLogger = l ?? undefined; }
  setStateLogger(l: ((e: StateLogEntry) => void) | null): void { this.stateLogger = l ?? undefined; }
  setErrorLogger(l: ((e: ErrorLogEntry) => void) | null): void { this.errorLogger = l ?? undefined; }
  removeAllLoggers(): void { this.transitionLogger = undefined; this.stateLogger = undefined; this.errorLogger = undefined; }

  async execute(initialData: Map<string, unknown>): Promise<FlowContext> {
    const flowId = crypto.randomUUID();
    const ctx = new FlowContext(flowId);
    for (const [k, v] of initialData) ctx.put(k as any, v);

    const completed: string[] = [];
    let prev = 'initial';

    for (const step of this.steps) {
      const stepStart = performance.now();
      this.transitionLogger?.({ flowId, flowName: this.name, from: prev, to: step.name, trigger: step.name, durationMicros: 0 });

      const keysBefore = this.stateLogger ? new Set(ctx.snapshot().keys()) : null;

      try {
        await step.process(ctx);
      } catch (e: any) {
        const err = e instanceof Error ? e : new Error(String(e));
        const durationMicros = Math.round((performance.now() - stepStart) * 1000);
        this.errorLogger?.({ flowId, flowName: this.name, from: prev, to: step.name, trigger: step.name, cause: err, durationMicros });
        throw new PipelineException(step.name, [...completed], ctx, err);
      }

      if (this.strictMode) {
        for (const prod of step.produces) {
          if (!ctx.has(prod)) {
            const err = new FlowError('PRODUCES_VIOLATION',
              `Step '${step.name}' declares produces ${prod} but did not put it`);
            throw new PipelineException(step.name, [...completed], ctx, err);
          }
        }
      }

      if (this.stateLogger && keysBefore) {
        for (const [k] of ctx.snapshot()) {
          if (!keysBefore.has(k)) {
            this.stateLogger({ flowId, flowName: this.name, state: step.name, key: k, value: ctx.snapshot().get(k) });
          }
        }
      }

      completed.push(step.name);
      prev = step.name;
    }
    return ctx;
  }

  asStep(): PipelineStep {
    const self = this;
    return {
      name: self.name,
      requires: [...self._initiallyAvailable] as FlowKey<unknown>[],
      produces: self.steps.flatMap(s => s.produces),
      async process(ctx: FlowContext) {
        for (const s of self.steps) await s.process(ctx);
      },
    };
  }

  static builder(name: string): PipelineBuilder { return new PipelineBuilder(name); }
}

export class PipelineBuilder {
  private readonly steps: PipelineStep[] = [];
  private readonly _initiallyAvailable = new Set<string>();

  constructor(private readonly name: string) {}

  initiallyAvailable(...keys: FlowKey<unknown>[]): this {
    for (const k of keys) this._initiallyAvailable.add(k as string);
    return this;
  }

  step(s: PipelineStep): this { this.steps.push(s); return this; }

  build(): Pipeline {
    const errors: string[] = [];
    const available = new Set(this._initiallyAvailable);
    for (const s of this.steps) {
      for (const req of s.requires) {
        if (!available.has(req as string))
          errors.push(`Step '${s.name}' requires ${req} but it may not be available`);
      }
      for (const p of s.produces) available.add(p as string);
    }
    if (errors.length > 0) {
      throw new FlowError('INVALID_PIPELINE',
        `Pipeline '${this.name}' has ${errors.length} error(s):\n  - ${errors.join('\n  - ')}`);
    }
    return (Pipeline as any).builder(this.name) as unknown as Pipeline;
    // Use private constructor via reflection-like pattern
  }
}

// Fix: expose Pipeline constructor to builder
Object.defineProperty(PipelineBuilder.prototype, 'build', {
  value: function(this: PipelineBuilder) {
    const errors: string[] = [];
    const available = new Set((this as any)._initiallyAvailable);
    for (const s of (this as any).steps) {
      for (const req of s.requires) {
        if (!available.has(req as string))
          errors.push(`Step '${s.name}' requires ${req} but it may not be available`);
      }
      for (const p of s.produces) available.add(p as string);
    }
    if (errors.length > 0) {
      throw new FlowError('INVALID_PIPELINE',
        `Pipeline '${(this as any).name}' has ${errors.length} error(s):\n  - ${errors.join('\n  - ')}`);
    }
    const pipeline = Object.create(Pipeline.prototype);
    Object.assign(pipeline, {
      name: (this as any).name,
      steps: [...(this as any).steps],
      _initiallyAvailable: new Set((this as any)._initiallyAvailable),
      _dataFlow: new PipelineDataFlow([...(this as any).steps], new Set((this as any)._initiallyAvailable)),
      strictMode: false,
    });
    return pipeline as Pipeline;
  },
});
